app/src/main/java/org/oppia/app/parser/StringToNumberWithUnitsParser.kt

```diff
@@ -0,0 +1,168 @@
+    var rawInput: String = inputText.normalizeWhitespace()
+    CURRENCY_UNITS = getCurrencyUnits().unitHashMap
+    rawInput = rawInput.trim()
+    var type = ""
+    var real = 0.0
+    // Default fraction value.
+    var fractionObj = Fraction.newBuilder().setDenominator(1).build()
+    var units = ""
+    var value = ""
+
+    // Allow validation only when rawInput is not null or an empty string.
+    if (rawInput.isNotEmpty() && rawInput != null) {
+      // Start with digit when there is no currency unit.
+      if (rawInput.matches("^\\-?\\d.*\$".toRegex())) {
+        var ind = indexOf(Pattern.compile("[a-z(₹$]"), rawInput)
+        if (ind == -1) {
+          // There is value with no units.
+          value = rawInput;
+          units = ""
+        } else {
+          ind -= 1
+          value = rawInput.substring(0, ind).trim()
+          units = rawInput.substring(ind).trim()
+        }
+
+        var keys = (CURRENCY_UNITS).keys
+        for (i in keys) {
+          for (j in 0 until CURRENCY_UNITS[i]!!.frontUnits.size) {
+            if (units.indexOf(CURRENCY_UNITS[i]!!.frontUnits[j]) != -1) {
+              throw  Error(
+                "INVALID_CURRENCY_FORMAT"
+              )
+            }
+          }
+        }
+      } else {
+        var startsWithCorrectCurrencyUnit = false
+        var keys = (CURRENCY_UNITS).keys
+        for (i in keys) {
+          for (j in 0 until CURRENCY_UNITS[i]!!.frontUnits.size) {
+            if (rawInput.startsWith(CURRENCY_UNITS[i]!!.frontUnits[j])) {
+              startsWithCorrectCurrencyUnit = true
+              break
+            }
+          }
+        }
+        if (startsWithCorrectCurrencyUnit == false) {
+          throw  Error("INVALID_CURRENCY")
+        }
+
+        var ind = indexOf(Pattern.compile("[0-9]"), rawInput)
+
+
+        if (ind === -1) {
+          throw  Error("INVALID_CURRENCY")
+        }
+        units = rawInput.substring(0, ind).trim()
+
+        startsWithCorrectCurrencyUnit = false
+        keys = (CURRENCY_UNITS).keys
+        for (i in keys) {
+          for (j in 0 until CURRENCY_UNITS[i]!!.frontUnits.size) {
+            if (units == (CURRENCY_UNITS[i]!!.frontUnits[j]).trim()) {
+              startsWithCorrectCurrencyUnit = true
+              break
+            }
+          }
+        }
+        if (startsWithCorrectCurrencyUnit == false) {
+          throw  Error("INVALID_CURRENCY")
+        }
+        units = units + ""
+
+        var ind2 = indexOf(Pattern.compile("[a-z(]"), rawInput.substring(ind))
+        if (ind2 != -1) {
+          value = rawInput.substring(ind, ind2 - ind).trim()
+          units += rawInput.substring(ind2).trim()
+        } else {
+          value = rawInput.substring(ind).trim()
+          units = units.trim()
+        }
+      }
+      // Checking invalid characters in value.
+      if (value.matches("[a - z]".toRegex()) || value.matches("[ *^$₹()#@]/".toRegex())) {
```

Comment:
Some of these patterns seem wrong to me--we shouldn't be hardcoding them. They need to be assembled based on the currencies that we've configured in the map. Can you update all relevant patterns to be based on the currency map instead of being hardcoded? Prefer having a single source of truth.

Reply:
PTAL on https://github.com/oppia/oppia/blob/410fc985080836d348aab101ca4a213a948ff713/core/templates/dev/head/domain/objects/NumberWithUnitsObjectFactory.ts#L220